source/grid/grid_tools.cc

```diff
@@ -1458,6 +1458,164 @@ namespace GridTools
+  {
+    std::pair<typename MeshType<dim, spacedim>::active_cell_iterator, Point<dim> > cell_and_position;
+
+    bool found_cell = false;
+
+    unsigned int closest_vertex_index = 0;
+    Tensor<1,spacedim> vertex_to_point;
+    auto current_cell = cell_hint;
+
+    while (found_cell == false)
+      {
+        // First look at the verticesl of the cell cell_hint. If it's an
+        // invalid cell, then query for the closest global vertex
+        if (current_cell.state() == IteratorState::valid)
+          {
+            const unsigned int closest_vertex = find_closest_vertex_of_cell<dim,spacedim>(current_cell , p);
+            vertex_to_point = p - current_cell ->vertex(closest_vertex);
+            closest_vertex_index = current_cell ->vertex_index(closest_vertex);
+          }
+        else
+          {
+            closest_vertex_index = GridTools::find_closest_vertex(mesh,p,marked_vertices);
+            vertex_to_point = p - mesh.get_vertices()[closest_vertex_index];
+          }
+
+        vertex_to_point /= vertex_to_point.norm();
+        const unsigned int n_neighbor_cells = vertex_to_cells[closest_vertex_index].size();
+
+        // Create a corresponding map of vectors from vertex to cell center
+        std::vector<unsigned int> neighbor_permutation(n_neighbor_cells);
+
+        for (unsigned int i=0; i<n_neighbor_cells; ++i)
+          neighbor_permutation[i] = i;
+
+        auto comp = [&](const unsigned int a, const unsigned int b) -> bool
+        {
+          return compare_point_association<spacedim>(a,b,vertex_to_point,vertex_to_cell_centers[closest_vertex_index]);
+        };
+
+        std::sort(neighbor_permutation.begin(),
+                  neighbor_permutation.end(),
+                  comp);
+
+        // Search all of the cells adjacent to the closest vertex of the cell hint
+        // Most likely we will find the point in them.
+        for (unsigned int i=0; i<n_neighbor_cells; ++i)
+          {
+            try
+              {
+                auto cell = vertex_to_cells[closest_vertex_index].begin();
+                std::advance(cell,neighbor_permutation[i]);
+                const Point<dim> p_unit = mapping.transform_real_to_unit_cell(*cell, p);
+                if (GeometryInfo<dim>::is_inside_unit_cell(p_unit))
+                  {
+                    cell_and_position.first = *cell;
+                    cell_and_position.second = p_unit;
+                    found_cell = true;
+                    break;
+                  }
+              }
+            catch (typename Mapping<dim>::ExcTransformationFailed &)
+              {}
+          }
+
+        // The first time around, we check for vertices in the hint_cell. If that
+        // does not work, we set the cell iterator to an invalid one, and look
+        // for a global vertex close to the point. If that does not work, we are in
+        // trouble, and just throw an exception.
+        AssertThrow((found_cell == false) ||
```

Comment:
this looks weird to me. Shouldn't it be ``found_cell == true || state==valid`` ? I think the logic would be easier if you just do ``if (found_cell) return cell_and_position;`` before this AssertThrow.